File: javascript/basic-usage.js
/**
 * Basic JavaScript/Node.js examples for AI Agent Passport Registry
 */

const https = require("https");

// Configuration
const API_BASE_URL = process.env.API_URL || "https://api.aport.io";
const ADMIN_TOKEN = process.env.ADMIN_TOKEN || "your-admin-token";

/**
 * Make HTTP request helper
 */
function makeRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const req = https.request(url, options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try {
          const jsonData = JSON.parse(data);
          resolve({
            status: res.statusCode,
            data: jsonData,
            headers: res.headers,
          });
        } catch (e) {
          resolve({ status: res.statusCode, data: data, headers: res.headers });
        }
      });
    });

    req.on("error", reject);

    if (options.body) {
      req.write(options.body);
    }

    req.end();
  });
}

/**
 * Verify an agent passport
 */
async function verifyPassport(agentId) {
  console.log(`\n🔍 Verifying passport for agent: ${agentId}`);

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/verify/${agentId}`);

    if (response.status === 200) {
      console.log("✅ Passport verified successfully:");
      console.log(JSON.stringify(response.data, null, 2));

      // Check rate limit headers
      console.log("\n📊 Rate Limit Info:");
      console.log(`Limit: ${response.headers["x-ratelimit-limit"]}`);
      console.log(`Remaining: ${response.headers["x-ratelimit-remaining"]}`);
      console.log(
        `Reset: ${new Date(
          parseInt(response.headers["x-ratelimit-reset"]) * 1000
        )}`
      );
    } else {
      console.log(`❌ Verification failed (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error("❌ Error verifying passport:", error.message);
  }
}

/**
 * Create a new agent passport
 */
async function createPassport(passportData) {
  console.log("\n📝 Creating new passport...");

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/admin/create`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${ADMIN_TOKEN}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(passportData),
    });

    if (response.status === 201) {
      console.log("✅ Passport created successfully:");
      console.log(JSON.stringify(response.data, null, 2));
    } else {
      console.log(`❌ Creation failed (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error("❌ Error creating passport:", error.message);
  }
}

/**
 * List all agents (admin only)
 */
async function listAgents() {
  console.log("\n📋 Listing all agents...");

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/admin/agents`, {
      headers: {
        Authorization: `Bearer ${ADMIN_TOKEN}`,
      },
    });

    if (response.status === 200) {
      console.log("✅ Agents retrieved successfully:");
      console.log(JSON.stringify(response.data, null, 2));
    } else {
      console.log(`❌ Failed to list agents (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error("❌ Error listing agents:", error.message);
  }
}

/**
 * Update agent status
 */
async function updateAgentStatus(agentId, status, reason = "") {
  console.log(`\n🔄 Updating agent ${agentId} status to ${status}...`);

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/admin/status`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${ADMIN_TOKEN}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        agent_id: agentId,
        status: status,
        reason: reason,
      }),
    });

    if (response.status === 200) {
      console.log("✅ Status updated successfully:");
      console.log(JSON.stringify(response.data, null, 2));
    } else {
      console.log(`❌ Status update failed (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error("❌ Error updating status:", error.message);
  }
}

/**
 * Get system metrics
 */
async function getMetrics() {
  console.log("\n📊 Getting system metrics...");

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/metrics`, {
      headers: {
        Authorization: `Bearer ${ADMIN_TOKEN}`,
      },
    });

    if (response.status === 200) {
      console.log("✅ Metrics retrieved successfully:");
      console.log(JSON.stringify(response.data, null, 2));
    } else {
      console.log(`❌ Failed to get metrics (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error("❌ Error getting metrics:", error.message);
  }
}

/**
 * Handle rate limiting with exponential backoff
 */
async function verifyWithRetry(agentId, maxRetries = 3) {
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      const response = await makeRequest(
        `${API_BASE_URL}/api/verify/${agentId}`
      );

      if (response.status === 200) {
        return response.data;
      } else if (response.status === 429) {
        // Rate limited
        const retryAfter = response.data.retryAfter || Math.pow(2, attempt);
        console.log(
          `⏳ Rate limited. Retrying in ${retryAfter} seconds... (attempt ${attempt}/${maxRetries})`
        );

        if (attempt < maxRetries) {
          await new Promise((resolve) =>
            setTimeout(resolve, retryAfter * 1000)
          );
          continue;
        }
      }

      throw new Error(`Request failed with status ${response.status}`);
    } catch (error) {
      if (attempt === maxRetries) {
        throw error;
      }
      console.log(`⚠️ Attempt ${attempt} failed: ${error.message}`);
    }
  }
}

/**
 * Demonstrate capabilities and limits enforcement
 */
async function demonstrateCapabilitiesAndLimits(agentId) {
  console.log(
    `\n🔐 Demonstrating capabilities and limits for agent: ${agentId}`
  );

  try {
    const response = await makeRequest(`${API_BASE_URL}/api/verify/${agentId}`);

    if (response.status === 200) {
      const passport = response.data;
      console.log("✅ Passport retrieved successfully");

      // Check capabilities
      console.log("\n📋 Capabilities:");
      if (passport.capabilities && passport.capabilities.length > 0) {
        passport.capabilities.forEach((cap) => {
          console.log(
            `  - ${cap.id}${
              cap.params ? ` (params: ${JSON.stringify(cap.params)})` : ""
            }`
          );
        });
      } else {
        console.log("  No capabilities defined");
      }

      // Check limits
      console.log("\n⚖️ Limits:");
      if (passport.limits) {
        Object.entries(passport.limits).forEach(([key, value]) => {
          console.log(`  - ${key}: ${value}`);
        });
      } else {
        console.log("  No limits defined");
      }

      // Demonstrate enforcement examples
      console.log("\n🛡️ Enforcement Examples:");

      // Refund capability check
      if (passport.capabilities?.some((cap) => cap.id === "payments.refund")) {
        console.log("  ✅ Agent has refund capability");

        // Check refund limits
        if (passport.limits?.refund_amount_max_per_tx) {
          const refundAmount = 5000; // $50.00 in cents
          if (refundAmount <= passport.limits.refund_amount_max_per_tx) {
            console.log(
              `  ✅ Refund amount $${
                refundAmount / 100
              } is within per-transaction limit of $${
                passport.limits.refund_amount_max_per_tx / 100
              }`
            );
          } else {
            console.log(
              `  ❌ Refund amount $${
                refundAmount / 100
              } exceeds per-transaction limit of $${
                passport.limits.refund_amount_max_per_tx / 100
              }`
            );
          }
        }
      } else {
        console.log("  ❌ Agent does not have refund capability");
      }

      // Data export capability check
      if (passport.capabilities?.some((cap) => cap.id === "data.export")) {
        console.log("  ✅ Agent has data export capability");

        // Check export limits
        if (passport.limits?.max_export_rows) {
          const requestedRows = 5000;
          if (requestedRows <= passport.limits.max_export_rows) {
            console.log(
              `  ✅ Export request for ${requestedRows} rows is within limit of ${passport.limits.max_export_rows}`
            );
          } else {
            console.log(
              `  ❌ Export request for ${requestedRows} rows exceeds limit of ${passport.limits.max_export_rows}`
            );
          }
        }

        // Check PII access
        if (passport.limits?.allow_pii !== undefined) {
          console.log(
            `  ${passport.limits.allow_pii ? "✅" : "❌"} PII access is ${
              passport.limits.allow_pii ? "allowed" : "not allowed"
            }`
          );
        }
      } else {
        console.log("  ❌ Agent does not have data export capability");
      }

      // Messaging capability check
      if (passport.capabilities?.some((cap) => cap.id === "messaging.send")) {
        console.log("  ✅ Agent has messaging capability");

        // Check message rate limits
        if (passport.limits?.msgs_per_min) {
          console.log(
            `  ✅ Message rate limit: ${passport.limits.msgs_per_min} per minute`
          );
        }
        if (passport.limits?.msgs_per_day) {
          console.log(
            `  ✅ Daily message limit: ${passport.limits.msgs_per_day} per day`
          );
        }

        // Check channel allowlist
        const messagingCap = passport.capabilities.find(
          (cap) => cap.id === "messaging.send"
        );
        if (messagingCap?.params?.channels_allowlist) {
          console.log(
            `  ✅ Allowed channels: ${messagingCap.params.channels_allowlist.join(
              ", "
            )}`
          );
        }
        if (messagingCap?.params?.mention_policy) {
          console.log(
            `  ✅ Mention policy: ${messagingCap.params.mention_policy}`
          );
        }
      } else {
        console.log("  ❌ Agent does not have messaging capability");
      }

      // Repository PR creation capability check
      if (passport.capabilities?.some((cap) => cap.id === "repo.pr.create")) {
        console.log("  ✅ Agent has PR creation capability");

        if (passport.limits?.max_prs_per_day) {
          console.log(
            `  ✅ Daily PR limit: ${passport.limits.max_prs_per_day} per day`
          );
        }

        const prCap = passport.capabilities.find(
          (cap) => cap.id === "repo.pr.create"
        );
        if (prCap?.params?.allowed_repos) {
          console.log(
            `  ✅ Allowed repositories: ${prCap.params.allowed_repos.join(
              ", "
            )}`
          );
        }
        if (prCap?.params?.allowed_base_branches) {
          console.log(
            `  ✅ Allowed base branches: ${prCap.params.allowed_base_branches.join(
              ", "
            )}`
          );
        }
      } else {
        console.log("  ❌ Agent does not have PR creation capability");
      }

      // Repository merge capability check
      if (passport.capabilities?.some((cap) => cap.id === "repo.merge")) {
        console.log("  ✅ Agent has merge capability");

        if (passport.limits?.max_merges_per_day) {
          console.log(
            `  ✅ Daily merge limit: ${passport.limits.max_merges_per_day} per day`
          );
        }
        if (passport.limits?.max_pr_size_kb) {
          console.log(`  ✅ Max PR size: ${passport.limits.max_pr_size_kb} KB`);
        }

        const mergeCap = passport.capabilities.find(
          (cap) => cap.id === "repo.merge"
        );
        if (mergeCap?.params?.required_reviews) {
          console.log(
            `  ✅ Required reviews: ${mergeCap.params.required_reviews}`
          );
        }
        if (mergeCap?.params?.required_labels) {
          console.log(
            `  ✅ Required labels: ${mergeCap.params.required_labels.join(
              ", "
            )}`
          );
        }
      } else {
        console.log("  ❌ Agent does not have merge capability");
      }

      // Assurance level check
      if (passport.assurance_level) {
        console.log(`\n🛡️ Assurance Level: ${passport.assurance_level}`);
        console.log(`  Method: ${passport.assurance_method || "N/A"}`);
        console.log(`  Verified: ${passport.assurance_verified_at || "N/A"}`);

        // Example assurance requirements
        const requiredLevels = {
          refunds: "L2",
          payouts: "L3",
          admin: "L4KYC",
        };

        console.log("\n🔒 Route Access Requirements:");
        Object.entries(requiredLevels).forEach(([route, requiredLevel]) => {
          const hasAccess = compareAssuranceLevels(
            passport.assurance_level,
            requiredLevel
          );
          console.log(
            `  ${route}: ${
              hasAccess ? "✅" : "❌"
            } (requires ${requiredLevel}, has ${passport.assurance_level})`
          );
        });
      }

      // Taxonomy information
      if (passport.categories || passport.framework) {
        console.log("\n🏷️ Taxonomy:");
        if (passport.categories && passport.categories.length > 0) {
          console.log(`  Categories: ${passport.categories.join(", ")}`);
        }
        if (passport.framework && passport.framework.length > 0) {
          console.log(`  Frameworks: ${passport.framework.join(", ")}`);
        }
      }
    } else {
      console.log(`❌ Failed to retrieve passport (${response.status}):`);
      console.log(JSON.stringify(response.data, null, 2));
    }
  } catch (error) {
    console.error(
      "❌ Error demonstrating capabilities and limits:",
      error.message
    );
  }
}

/**
 * Simple assurance level comparison helper
 */
function compareAssuranceLevels(current, required) {
  const levels = ["L0", "L1", "L2", "L3", "L4KYC", "L4FIN"];
  const currentIndex = levels.indexOf(current);
  const requiredIndex = levels.indexOf(required);
  return currentIndex >= requiredIndex;
}

// Example usage
async function main() {
  console.log("🚀 AI Agent Passport Registry - JavaScript Examples\n");

  // Verify existing passports
  await verifyPassport("ap_demo_001");
  await verifyPassport("ap_128094d3");

  // Demonstrate capabilities and limits enforcement
  await demonstrateCapabilitiesAndLimits("ap_demo_001");

  // Create a new passport
  const newPassport = {
    agent_id: "ap_js_example",
    owner: "JavaScript Example",
    role: "Tier-1",
    permissions: ["read:data", "create:reports"],
    limits: {
      api_calls_per_hour: 500,
      ticket_creation_daily: 25,
    },
    regions: ["US-CA"],
    status: "active",
    contact: "example@javascript.com",
    version: "1.0.0",
  };

  await createPassport(newPassport);

  // Create a passport with new capabilities
  const newCapabilitiesPassport = {
    agent_id: "ap_js_new_caps",
    owner: "JavaScript New Capabilities Example",
    role: "agent",
    capabilities: [
      {
        id: "messaging.send",
        params: {
          channels_allowlist: ["slack", "discord", "email"],
          mention_policy: "limited",
        },
      },
      {
        id: "repo.pr.create",
        params: {
          allowed_repos: ["company/public-repo", "company/docs"],
          allowed_base_branches: ["main", "develop"],
          path_allowlist: ["src/**", "docs/**"],
          max_files_changed: 20,
          max_total_added_lines: 500,
        },
      },
      {
        id: "repo.merge",
        params: {
          allowed_repos: ["company/public-repo"],
          allowed_base_branches: ["develop"],
          required_labels: ["approved", "tested"],
          required_reviews: 2,
        },
      },
    ],
    limits: {
      msgs_per_min: 30,
      msgs_per_day: 1000,
      max_prs_per_day: 10,
      max_merges_per_day: 5,
      max_pr_size_kb: 512,
    },
    regions: ["global"],
    status: "active",
    contact: "newcaps@javascript.com",
    version: "1.0.0",
  };

  await createPassport(newCapabilitiesPassport);

  // List all agents
  await listAgents();

  // Update agent status
  await updateAgentStatus("ap_js_example", "suspended", "Testing suspension");

  // Get metrics
  await getMetrics();

  // Example with rate limiting
  console.log("\n🔄 Testing rate limiting with retry...");
  try {
    const result = await verifyWithRetry("ap_demo_001");
    console.log("✅ Verification with retry successful:", result);
  } catch (error) {
    console.log("❌ Verification with retry failed:", error.message);
  }

  console.log("\n✨ Examples completed!");
}

// Run examples if this file is executed directly
if (require.main === module) {
  main().catch(console.error);
}

module.exports = {
  verifyPassport,
  createPassport,
  listAgents,
  updateAgentStatus,
  getMetrics,
  verifyWithRetry,
};

File: python/basic_usage.py
"""
Basic Python examples for AI Agent Passport Registry
"""

import requests
import time
import json
import os
from typing import Dict, Any, Optional

# Configuration
API_BASE_URL = os.getenv('API_URL', 'https://api.aport.io')
ADMIN_TOKEN = os.getenv('ADMIN_TOKEN', 'your-admin-token')

class AgentPassportClient:
    """Client for interacting with the AI Agent Passport Registry API"""
    
    def __init__(self, base_url: str = API_BASE_URL, admin_token: str = ADMIN_TOKEN):
        self.base_url = base_url
        self.admin_token = admin_token
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            'User-Agent': 'Python-Client/1.0',
            'Accept': 'application/json'
        })
    
    def _make_request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """Make HTTP request with error handling"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            response = self.session.request(method, url, **kwargs)
            
            # Try to parse JSON response
            try:
                data = response.json()
            except ValueError:
                data = response.text
            
            return {
                'status_code': response.status_code,
                'data': data,
                'headers': dict(response.headers)
            }
        except requests.RequestException as e:
            return {
                'status_code': 0,
                'data': {'error': str(e)},
                'headers': {}
            }
    
    def verify_passport(self, agent_id: str) -> Dict[str, Any]:
        """Verify an agent passport"""
        print(f"\n🔍 Verifying passport for agent: {agent_id}")
        
        response = self._make_request('GET', f'/api/verify/{agent_id}')
        
        if response['status_code'] == 200:
            print('✅ Passport verified successfully:')
            print(json.dumps(response['data'], indent=2))
            
            # Check rate limit headers
            headers = response['headers']
            print('\n📊 Rate Limit Info:')
            print(f"Limit: {headers.get('X-RateLimit-Limit', 'N/A')}")
            print(f"Remaining: {headers.get('X-RateLimit-Remaining', 'N/A')}")
            print(f"Reset: {headers.get('X-RateLimit-Reset', 'N/A')}")
        else:
            print(f"❌ Verification failed ({response['status_code']}):")
            print(json.dumps(response['data'], indent=2))
        
        return response
    
    def create_passport(self, passport_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new agent passport"""
        print('\n📝 Creating new passport...')
        
        response = self._make_request(
            'POST',
            '/api/admin/create',
            headers={'Authorization': f'Bearer {self.admin_token}'},
            json=passport_data
        )
        
        if response['status_code'] == 201:
            print('✅ Passport created successfully:')
            print(json.dumps(response['data'], indent=2))
        else:
            print(f"❌ Creation failed ({response['status_code']}):")
            print(json.dumps(response['data'], indent=2))
        
        return response
    
    def list_agents(self) -> Dict[str, Any]:
        """List all agents (admin only)"""
        print('\n📋 Listing all agents...')
        
        response = self._make_request(
            'GET',
            '/api/admin/agents',
            headers={'Authorization': f'Bearer {self.admin_token}'}
        )
        
        if response['status_code'] == 200:
            print('✅ Agents retrieved successfully:')
            print(json.dumps(response['data'], indent=2))
        else:
            print(f"❌ Failed to list agents ({response['status_code']}):")
            print(json.dumps(response['data'], indent=2))
        
        return response
    
    def update_agent_status(self, agent_id: str, status: str, reason: str = '') -> Dict[str, Any]:
        """Update agent status"""
        print(f'\n🔄 Updating agent {agent_id} status to {status}...')
        
        response = self._make_request(
            'POST',
            '/api/admin/status',
            headers={'Authorization': f'Bearer {self.admin_token}'},
            json={
                'agent_id': agent_id,
                'status': status,
                'reason': reason
            }
        )
        
        if response['status_code'] == 200:
            print('✅ Status updated successfully:')
            print(json.dumps(response['data'], indent=2))
        else:
            print(f"❌ Status update failed ({response['status_code']}):")
            print(json.dumps(response['data'], indent=2))
        
        return response
    
    def demonstrate_capabilities_and_limits(self, agent_id: str) -> Dict[str, Any]:
        """Demonstrate capabilities and limits enforcement"""
        print(f'\n🔐 Demonstrating capabilities and limits for agent: {agent_id}')
        
        response = self.verify_passport(agent_id)
        
        if response['status_code'] == 200:
            passport = response['data']
            print('✅ Passport retrieved successfully')
            
            # Check capabilities
            print('\n📋 Capabilities:')
            if passport.get('capabilities') and len(passport['capabilities']) > 0:
                for cap in passport['capabilities']:
                    params_str = f" (params: {json.dumps(cap.get('params', {}))})" if cap.get('params') else ''
                    print(f"  - {cap['id']}{params_str}")
            else:
                print('  No capabilities defined')
            
            # Check limits
            print('\n⚖️ Limits:')
            if passport.get('limits'):
                for key, value in passport['limits'].items():
                    print(f'  - {key}: {value}')
            else:
                print('  No limits defined')
            
            # Demonstrate enforcement examples
            print('\n🛡️ Enforcement Examples:')
            
            # Refund capability check
            has_refund_cap = any(cap['id'] == 'payments.refund' for cap in passport.get('capabilities', []))
            if has_refund_cap:
                print('  ✅ Agent has refund capability')
                
                # Check refund limits
                if passport.get('limits', {}).get('refund_amount_max_per_tx'):
                    refund_amount = 5000  # $50.00 in cents
                    max_per_tx = passport['limits']['refund_amount_max_per_tx']
                    if refund_amount <= max_per_tx:
                        print(f'  ✅ Refund amount ${refund_amount/100} is within per-transaction limit of ${max_per_tx/100}')
                    else:
                        print(f'  ❌ Refund amount ${refund_amount/100} exceeds per-transaction limit of ${max_per_tx/100}')
            else:
                print('  ❌ Agent does not have refund capability')
            
            # Data export capability check
            has_export_cap = any(cap['id'] == 'data.export' for cap in passport.get('capabilities', []))
            if has_export_cap:
                print('  ✅ Agent has data export capability')
                
                # Check export limits
                if passport.get('limits', {}).get('max_export_rows'):
                    requested_rows = 5000
                    max_rows = passport['limits']['max_export_rows']
                    if requested_rows <= max_rows:
                        print(f'  ✅ Export request for {requested_rows} rows is within limit of {max_rows}')
                    else:
                        print(f'  ❌ Export request for {requested_rows} rows exceeds limit of {max_rows}')
                
                # Check PII access
                allow_pii = passport.get('limits', {}).get('allow_pii')
                if allow_pii is not None:
                    status = '✅' if allow_pii else '❌'
                    access = 'allowed' if allow_pii else 'not allowed'
                    print(f'  {status} PII access is {access}')
            else:
                print('  ❌ Agent does not have data export capability')
            
            # Messaging capability check
            has_messaging_cap = any(cap['id'] == 'messaging.send' for cap in passport.get('capabilities', []))
            if has_messaging_cap:
                print('  ✅ Agent has messaging capability')
                
                # Check message rate limits
                if passport.get('limits', {}).get('msgs_per_min'):
                    print(f'  ✅ Message rate limit: {passport["limits"]["msgs_per_min"]} per minute')
                if passport.get('limits', {}).get('msgs_per_day'):
                    print(f'  ✅ Daily message limit: {passport["limits"]["msgs_per_day"]} per day')
                
                # Check channel allowlist
                messaging_cap = next((cap for cap in passport.get('capabilities', []) if cap['id'] == 'messaging.send'), None)
                if messaging_cap and messaging_cap.get('params', {}).get('channels_allowlist'):
                    channels = ', '.join(messaging_cap['params']['channels_allowlist'])
                    print(f'  ✅ Allowed channels: {channels}')
                if messaging_cap and messaging_cap.get('params', {}).get('mention_policy'):
                    print(f'  ✅ Mention policy: {messaging_cap["params"]["mention_policy"]}')
            else:
                print('  ❌ Agent does not have messaging capability')
            
            # Repository PR creation capability check
            has_pr_cap = any(cap['id'] == 'repo.pr.create' for cap in passport.get('capabilities', []))
            if has_pr_cap:
                print('  ✅ Agent has PR creation capability')
                
                if passport.get('limits', {}).get('max_prs_per_day'):
                    print(f'  ✅ Daily PR limit: {passport["limits"]["max_prs_per_day"]} per day')
                
                pr_cap = next((cap for cap in passport.get('capabilities', []) if cap['id'] == 'repo.pr.create'), None)
                if pr_cap and pr_cap.get('params', {}).get('allowed_repos'):
                    repos = ', '.join(pr_cap['params']['allowed_repos'])
                    print(f'  ✅ Allowed repositories: {repos}')
                if pr_cap and pr_cap.get('params', {}).get('allowed_base_branches'):
                    branches = ', '.join(pr_cap['params']['allowed_base_branches'])
                    print(f'  ✅ Allowed base branches: {branches}')
            else:
                print('  ❌ Agent does not have PR creation capability')
            
            # Repository merge capability check
            has_merge_cap = any(cap['id'] == 'repo.merge' for cap in passport.get('capabilities', []))
            if has_merge_cap:
                print('  ✅ Agent has merge capability')
                
                if passport.get('limits', {}).get('max_merges_per_day'):
                    print(f'  ✅ Daily merge limit: {passport["limits"]["max_merges_per_day"]} per day')
                if passport.get('limits', {}).get('max_pr_size_kb'):
                    print(f'  ✅ Max PR size: {passport["limits"]["max_pr_size_kb"]} KB')
                
                merge_cap = next((cap for cap in passport.get('capabilities', []) if cap['id'] == 'repo.merge'), None)
                if merge_cap and merge_cap.get('params', {}).get('required_reviews'):
                    print(f'  ✅ Required reviews: {merge_cap["params"]["required_reviews"]}')
                if merge_cap and merge_cap.get('params', {}).get('required_labels'):
                    labels = ', '.join(merge_cap['params']['required_labels'])
                    print(f'  ✅ Required labels: {labels}')
            else:
                print('  ❌ Agent does not have merge capability')
            
            # Assurance level check
            if passport.get('assurance_level'):
                print(f'\n🛡️ Assurance Level: {passport["assurance_level"]}')
                print(f'  Method: {passport.get("assurance_method", "N/A")}')
                print(f'  Verified: {passport.get("assurance_verified_at", "N/A")}')
                
                # Example assurance requirements
                required_levels = {
                    'refunds': 'L2',
                    'payouts': 'L3',
                    'admin': 'L4KYC'
                }
                
                print('\n🔒 Route Access Requirements:')
                for route, required_level in required_levels.items():
                    has_access = self._compare_assurance_levels(passport['assurance_level'], required_level)
                    status = '✅' if has_access else '❌'
                    print(f'  {route}: {status} (requires {required_level}, has {passport["assurance_level"]})')
            
            # Taxonomy information
            if passport.get('categories') or passport.get('framework'):
                print('\n🏷️ Taxonomy:')
                if passport.get('categories'):
                    print(f'  Categories: {", ".join(passport["categories"])}')
                if passport.get('framework'):
                    print(f'  Frameworks: {", ".join(passport["framework"])}')
        
        return response
    
    def _compare_assurance_levels(self, current: str, required: str) -> bool:
        """Simple assurance level comparison helper"""
        levels = ['L0', 'L1', 'L2', 'L3', 'L4KYC', 'L4FIN']
        try:
            current_index = levels.index(current)
            required_index = levels.index(required)
            return current_index >= required_index
        except ValueError:
            return False
    
    def get_metrics(self) -> Dict[str, Any]:
        """Get system metrics"""
        print('\n📊 Getting system metrics...')
        
        response = self._make_request(
            'GET',
            '/api/metrics',
            headers={'Authorization': f'Bearer {self.admin_token}'}
        )
        
        if response['status_code'] == 200:
            print('✅ Metrics retrieved successfully:')
            print(json.dumps(response['data'], indent=2))
        else:
            print(f"❌ Failed to get metrics ({response['status_code']}):")
            print(json.dumps(response['data'], indent=2))
        
        return response
    
    def verify_with_retry(self, agent_id: str, max_retries: int = 3) -> Optional[Dict[str, Any]]:
        """Verify passport with exponential backoff retry"""
        for attempt in range(1, max_retries + 1):
            try:
                response = self.verify_passport(agent_id)
                
                if response['status_code'] == 200:
                    return response['data']
                elif response['status_code'] == 429:
                    # Rate limited
                    retry_after = response['data'].get('retryAfter', 2 ** attempt)
                    print(f"⏳ Rate limited. Retrying in {retry_after} seconds... (attempt {attempt}/{max_retries})")
                    
                    if attempt < max_retries:
                        time.sleep(retry_after)
                        continue
                else:
                    print(f"❌ Request failed with status {response['status_code']}")
                    if attempt < max_retries:
                        time.sleep(2 ** attempt)
                        continue
                
            except Exception as e:
                print(f"⚠️ Attempt {attempt} failed: {str(e)}")
                if attempt < max_retries:
                    time.sleep(2 ** attempt)
                    continue
        
        return None

def main():
    """Run example usage"""
    print('🚀 AI Agent Passport Registry - Python Examples\n')
    
    # Initialize client
    client = AgentPassportClient()
    
    # Verify existing passports
    client.verify_passport('ap_demo_001')
    client.verify_passport('ap_128094d3')

    # Demonstrate capabilities and limits enforcement
    client.demonstrate_capabilities_and_limits('ap_demo_001')
    
    # Create a new passport
    new_passport = {
        'agent_id': 'ap_python_example',
        'owner': 'Python Example',
        'role': 'Tier-1',
        'permissions': ['read:data', 'create:reports'],
        'limits': {
            'api_calls_per_hour': 500,
            'ticket_creation_daily': 25
        },
        'regions': ['US-CA'],
        'status': 'active',
        'contact': 'example@python.com',
        'version': '1.0.0'
    }
    
    client.create_passport(new_passport)
    
    # Create a passport with new capabilities
    new_capabilities_passport = {
        'agent_id': 'ap_python_new_caps',
        'owner': 'Python New Capabilities Example',
        'role': 'agent',
        'capabilities': [
            {
                'id': 'messaging.send',
                'params': {
                    'channels_allowlist': ['slack', 'discord', 'email'],
                    'mention_policy': 'limited'
                }
            },
            {
                'id': 'repo.pr.create',
                'params': {
                    'allowed_repos': ['company/public-repo', 'company/docs'],
                    'allowed_base_branches': ['main', 'develop'],
                    'path_allowlist': ['src/**', 'docs/**'],
                    'max_files_changed': 20,
                    'max_total_added_lines': 500
                }
            },
            {
                'id': 'repo.merge',
                'params': {
                    'allowed_repos': ['company/public-repo'],
                    'allowed_base_branches': ['develop'],
                    'required_labels': ['approved', 'tested'],
                    'required_reviews': 2
                }
            }
        ],
        'limits': {
            'msgs_per_min': 30,
            'msgs_per_day': 1000,
            'max_prs_per_day': 10,
            'max_merges_per_day': 5,
            'max_pr_size_kb': 512
        },
        'regions': ['global'],
        'status': 'active',
        'contact': 'newcaps@python.com',
        'version': '1.0.0'
    }
    
    client.create_passport(new_capabilities_passport)
    
    # List all agents
    client.list_agents()
    
    # Update agent status
    client.update_agent_status('ap_python_example', 'suspended', 'Testing suspension')
    
    # Get metrics
    client.get_metrics()
    
    # Example with rate limiting
    print('\n🔄 Testing rate limiting with retry...')
    result = client.verify_with_retry('ap_demo_001')
    if result:
        print('✅ Verification with retry successful:', result)
    else:
        print('❌ Verification with retry failed')
    
    print('\n✨ Examples completed!')

if __name__ == '__main__':
    main()
